src/pipdeptree/_models/package.py

```diff
@@ -118,31 +110,73 @@ class DistPackage(Package):
 
     """
 
-    def __init__(self, obj: DistInfoDistribution, req: ReqPackage | None = None) -> None:
-        super().__init__(obj)
+    def __init__(self, obj: Distribution, req: ReqPackage | None = None) -> None:
+        super().__init__(obj.metadata["Name"])
+        self._obj = obj
         self.req = req
-        self._project_name = ""
 
     def requires(self) -> list[Requirement]:
-        return self._obj.requires()  # type: ignore[no-untyped-call,no-any-return]
+        req_list = []
+        req_name_list = []
+        if self._obj.requires:
+            for r in self._obj.requires:
+                req = Requirement(r)
+                is_extra_req = req.marker and contains_extra(str(req.marker))
+                if not is_extra_req and req.name not in req_name_list:
+                    req_list.append(req)
+                    req_name_list.append(req.name)
+        return req_list
 
     @property
-    def project_name(self) -> str:
-        if not self._project_name:
-            try:
-                self._project_name = metadata(self.key)["name"]
-            except (PackageNotFoundError, KeyError):
-                self._project_name = self._obj.project_name
-        return self._project_name
+    def editable(self) -> bool:
+        return self.direct_url_dict["editable"]
+
+    @property
+    def direct_url(self) -> str:
+        return self.direct_url_dict["url"]
+
+    @property
+    def raw_name(self) -> str:
+        return self.project_name
 
     @property
     def version(self) -> str:
         return self._obj.version  # type: ignore[no-any-return]
 
+    @property
+    def editable_project_location(self) -> str:
+        if self.direct_url:
+            from pip._internal.utils.urls import url_to_path  # noqa: PLC2701, PLC0415
+
+            return url_to_path(self.direct_url)
+
+        egg_link_path = egg_link_path_from_sys_path(self.raw_name)
+        if egg_link_path:
+            return self.location
+        return None
+
+    @property
+    def direct_url_dict(self) -> dict[str, Any]:
+        result = {"editable": False, "url": None}
+
+        if not self._obj.files:
+            return result
+
+        for path in self._obj.files:
+            if Path(path).name == "direct_url.json":
+                abstract_path = Path(self._obj.locate_file(path))
+                with abstract_path.open("r") as f:
+                    drurl = DirectUrl.from_json(f.read())
+                    result["url"] = drurl.url
+                    result["editable"] = bool(drurl.is_local_editable)
+                break
```

Comment:
This is expensive to perform (we are iterating a list of files and more importantly if we find the direct URL we would be parsing JSON). If you look at `editable_project_location()`, we use `self.direct_url`, which would mean that we call `direct_url_dict()` twice. We should only call it once and cache the results later (and only if we need it; remember, this is only used when generating frozen requirements i.e. `pipdeptree -f`).

Reply:
We can use `read_text()` method to read the `direct_url.json` like [here](https://github.com/pypa/pip/blob/f5e4ee104e7b171a7cfb2843c9c602abf7a4e346/src/pip/_internal/metadata/base.py#L296)